tls/s2n_config.h

```diff
@@ -39,6 +40,10 @@ struct s2n_config {
     unsigned disable_x509_validation:1;
     unsigned max_verify_cert_chain_depth_set:1;
 
+    /* list of groups to generate keyshare */
+    struct s2n_array *preferred_key_shares;
```

Comment:
Couldn't this just be a bitfield of enabled key share? That makes it clear that order is still handled by s2n, and avoids allocating space on the heap. If you go that route, I would also suggest changing the "client_send_empty_key_shares" flag to something like "send_custom_key_share_list". Something that indicates whether or not to use the bitfield.